mitiq/observable/observable.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import copy
+from typing import cast, FrozenSet, List, Set
+
+import numpy as np
+import cirq
+
+from mitiq.observable.pauli import PauliString
+
+
+class Observable:
+    def __init__(self, *paulis: PauliString) -> None:
+        self._paulis = set(paulis)
+        self._nterms = len(self._paulis)
+
+    @property
+    def nterms(self) -> int:
+        return self._nterms
+
+    def _qubits(self) -> Set[cirq.Qid]:
+        """Returns all qubits acted on by the Observable."""
+        return {q for pauli in self._paulis for q in pauli._pauli.qubits}
+
+    @property
+    def qubit_indices(self) -> List[int]:
+        return [cast(cirq.LineQubit, q).x for q in sorted(self._qubits())]
+
+    @property
+    def nqubits(self) -> int:
+        return len(self.qubit_indices)
+
+    def partition(self) -> Set[FrozenSet[PauliString]]:
+        plists: List[List[PauliString]] = []
+        paulis = copy.deepcopy(self._paulis)
+
+        while paulis:
+            pauli = paulis.pop()
+            added = False
+            for (i, plist) in enumerate(plists):
+                if all(pauli.can_be_measured_with(p) for p in plist):
+                    plists[i].append(pauli)
+                    added = True
+                    break
+
+            if not added:
+                plists.append([pauli])
+
+        return set([frozenset(plist) for plist in plists])
```

Comment:
What happens if we just return `plists`?

Is `frozenset` necessary for something? 
Is the final `set()` conversion necessary? 

Note: considering that `masure_in` returns a list of circuits, it could be more natural if partition returned a list too.

Reply:
Agree. I will change the return type to `List[Frozenset[PauliString]]`. This is actually essential for estimating the expectation value of the observable - you get a bunch of circuits to run then you get the measurements for each circuit, and you need them to match, but set is unordered.

FrozenSet was needed (as opposed to set) because the objects in a set need to be hashable. I think we should still keep it when changing to a list, though. 

P.S. I considered defining another class for a set of `PauliString`s which can be measured at the same time via single-qubit basis rotations. This would take the place of `FrozenSet[PauliString]`. It may be a good idea, but for now I'm just sticking to `FrozenSet[PauliString]`.

Reply:
Actually changed to `List[List[PauliString]]` due to note about duplicates. I almost want to consider something like `Observable(PauliString("Z"), PauliString("Z"))` instead of 2 * Z user error, but it should not change the PauliStrings in such a way that they become unequal.